CodeGen: Add ISD::AssertNoFPClass

wzssyqa opened the pull request:
It is used to mark a value that we are sure that it is not some fcType.
The examples include:
  * An arguments of a function is marked with nofpclass
  * Output value of an intrinsic can be sure to not be some type

So that the following operation can make some assumptions.

wzssyqa committed:
CodeGen: Add ISD::AssertNoFPClass

It is used to mark a value that we are sure that it is not some fcType.
The examples include:
  * An arguments of a function is marked with nofpclass
  * Output value of an intrinsic can be sure to not be some type
So that the following operation can make some assumptions.

wzssyqa commented:
One example about how to use it: https://github.com/llvm/llvm-project/pull/130051

arsenm commented on llvm/include/llvm/Target/TargetSelectionDAG.td:
```diff
@@ -859,6 +859,7 @@ def SDT_assert : SDTypeProfile<1, 1,
   [SDTCisInt<0>, SDTCisInt<1>, SDTCisSameAs<1, 0>]>;
 def assertsext : SDNode<"ISD::AssertSext", SDT_assert>;
 def assertzext : SDNode<"ISD::AssertZext", SDT_assert>;
+def assernofpclass : SDNode<"ISD::AssertNoFPClass", SDTFPUnaryOp>;
```
```suggestion
def assertnofpclass : SDNode<"ISD::AssertNoFPClass", SDTFPUnaryOp>;
```

arsenm reviewed (commented):
This patch should include the wiring up to the argument lowering with tests

wzssyqa committed:
fix typo assernofpclass

wzssyqa committed:
Add isKnownNeverNaN test

wzssyqa committed:
Mention operands

wzssyqa committed:
Fix typo

wzssyqa committed:
Fix AssertNoFPClass assert

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -5760,6 +5760,17 @@ bool SelectionDAG::isKnownNeverNaN(SDValue Op, const APInt &DemandedElts,
     return true;
   }
+  case ISD::AssertNoFPClass: {
+    SDValue SDNoFPClass = Op.getOperand(1);
+    assert(isa<ConstantSDNode>(SDNoFPClass) && "NoFPClass is not Constant");
+    FPClassTest NoFPClass = static_cast<FPClassTest>(
+        dyn_cast<ConstantSDNode>(SDNoFPClass)->getZExtValue());
```
```suggestion
    FPClassTest NoFPClass = static_cast<FPClassTest>(Op.getConstantOperandVal(1));
```

cast instead of isa + dyn_cast, but here we have a helper

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -5760,6 +5760,17 @@ bool SelectionDAG::isKnownNeverNaN(SDValue Op, const APInt &DemandedElts,
+    assert(isa<ConstantSDNode>(SDNoFPClass) && "NoFPClass is not Constant");
+    FPClassTest NoFPClass = static_cast<FPClassTest>(
+        dyn_cast<ConstantSDNode>(SDNoFPClass)->getZExtValue());
+    if (NoFPClass & fcNan)
+      return true;
```
```suggestion
    if ((NoFPClass & fcNan) == fcNan)
      return true;
```

Need to defend against the case where only one of the bits is set

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -7406,6 +7417,17 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
+  case ISD::AssertNoFPClass: {
+    assert(N1.getValueType().isFloatingPoint() &&
+           "AssertNoFPClass is used for a non-floating type");
+    assert(isa<ConstantSDNode>(N2) && "NoFPClass is not Constant");
+    FPClassTest NoFPClass =
+        static_cast<FPClassTest>(dyn_cast<ConstantSDNode>(N2)->getZExtValue());
+    assert(llvm::to_underlying(NoFPClass) <=
+               BitmaskEnumDetail::Mask<FPClassTest>() &&
+           "FPClassTest value too large");
```
Use getConstantOperandVal like above

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -11848,6 +11848,16 @@ void SelectionDAGISel::LowerArguments(const Function &F) {
+    if (NoFPClass != fcNone) {
+      EVT I64EVT = EVT::getIntegerVT(*DAG.getContext(), 64);
+      SDValue SDNoFPClass =
+          DAG.getConstant(static_cast<uint64_t>(NoFPClass), dl, I64EVT);
```
Should be getTargetConstant

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -11848,6 +11848,16 @@ void SelectionDAGISel::LowerArguments(const Function &F) {
+      EVT I64EVT = EVT::getIntegerVT(*DAG.getContext(), 64);
+      SDValue SDNoFPClass =
+          DAG.getConstant(static_cast<uint64_t>(NoFPClass), dl, I64EVT);
+      SDNodeFlags ResFlags = Res->getFlags();
```
Don't propagate the flags

arsenm commented on llvm/test/CodeGen/AArch64/nofpclass.ll:
```diff
@@ -0,0 +1,12 @@
+entry:
+  %cond = tail call float @llvm.maximumnum.f32(float %a, float %b)
+  ret float %cond
+}
```
Also test a vector case. And what about uniform float aggregates?

Commit:
Fix code styles

1. Use Op.getConstantOperandVal to get Constant value
2. (NoFPClass & fcNan) == fcNan
3. Use N2->getAsZExtValue() to get Constant value
4. Use getTargetConstant with ptr width to fix 32bit target
5. Don't propagate the flags
6. Add MIPSr6 testcase for 32bit target tests

wzssyqa committed:
Add Vector test case

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -11850,12 +11850,11 @@ void SelectionDAGISel::LowerArguments(const Function &F) {
-          DAG.getConstant(static_cast<uint64_t>(NoFPClass), dl, I64EVT);
-      SDNodeFlags ResFlags = Res->getFlags();
+          DAG.getTargetConstant(static_cast<uint64_t>(NoFPClass), dl,
+                                TLI->getPointerTy(DAG.getDataLayout()));
```
This should not use getPointerTy. I would make it an unconditional i32

wzssyqa replied:
Maybe a good idea, since no 16bit target is supported by LLVM yet, while I wish that we can keep the possibility.

arsenm replied:
The type of target constants does not matter at all. Making this a variable type is added complexity for no benefit

arsenm commented on llvm/test/CodeGen/AArch64/nofpclass.ll:
```diff
@@ -10,3 +10,13 @@ entry:
+entry:
+  %c = call <4 x float> @llvm.maximumnum.v4f32(<4 x float> %a, <4 x float> %b)
+  ret <4 x float> %c
+}
```
also test aggregate case, e.g. 

```
define { float, float } @struct({ float, float } nofpclass(nan) %a) {
   ret {float, float} %a
 }
```

RKSimon commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -5761,13 +5761,11 @@ bool SelectionDAG::isKnownNeverNaN(SDValue Op, const APInt &DemandedElts,
-    if (SNaN && (NoFPClass & fcSNan))
+    if (SNaN && (NoFPClass & fcSNan) == fcSNan)
       return true;
     return isKnownNeverNaN(Op.getOperand(0), SNaN, Depth + 1);
```
Add DemandedElts?

wzssyqa committed:
Add DemandedElts

wzssyqa committed:
Use 32bit target

wzssyqa committed:
Fix MERGE_VALUE cases

arsenm commented on llvm/test/CodeGen/AArch64/nofpclass.ll:
```diff
@@ -0,0 +1,29 @@
+}
+
+define { float, float } @struct({ float, float } nofpclass(nan) %a) {
+; CHECK-LABEL: struct:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    ret
```
Extract a component and test the result like in the other cases?

wzssyqa committed:
Add aggregate tests with fmin

wzssyqa committed:
Fix comment

wzssyqa committed:
Add snan and qnan test cases

arsenm commented on llvm/test/CodeGen/AArch64/nofpclass.ll:
```diff
@@ -21,9 +21,28 @@ entry:
    ret {float, float} %a
 }
+
+%struct.f2 = type { float, float }
```
Test the array case, but what I was asking was to test this struct type as the input

arsenm commented on llvm/test/CodeGen/Mips/nofpclass.ll:
```diff
@@ -16,3 +16,45 @@ entry:
 }
+
+define { float, float } @struct({ float, float} nofpclass(nan) %a) {
+; MIPS32R6-LABEL: struct:
+; MIPS32R6:       # %bb.0:
+; MIPS32R6-NEXT:    mov.s $f2, $f14
+; MIPS32R6-NEXT:    jr $ra
+; MIPS32R6-NEXT:    mov.s $f0, $f12
+;
+; MIPS64R6-LABEL: struct:
+; MIPS64R6:       # %bb.0:
+; MIPS64R6-NEXT:    mov.s $f2, $f13
+; MIPS64R6-NEXT:    jr $ra
+; MIPS64R6-NEXT:    mov.s $f0, $f12
```
Missing a use that shows this did something

wzssyqa replied:
Let's remove them since we have some new tests with max.

wzssyqa committed:
Test both array and struct

john-brawn-arm commented:
It looks like there's a different failure with the following:
```
target datalayout = "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64"
target triple = "thumbv7em-unknown-none-eabi"

define dso_local noundef nofpclass(nan inf) half @fn(half noundef returned nofpclass(nan inf) %x) {
entry:
  ret half %x
}
```
`llc -debug` gives
```
Legalizing node: t6: f16 = AssertNoFPClass t4, TargetConstant:i32<519>
Analyzing result type: f16
Soft promote half result 0: t6: f16 = AssertNoFPClass t4, TargetConstant:i32<519>
SoftPromoteHalfResult #0: t6: f16 = AssertNoFPClass t4, TargetConstant:i32<519>

LLVM ERROR: Do not know how to soft promote this operator's result!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /data_nvme1n1/johbra01/llvm-project/build/bin/llc tmp.ll -o - -debug
1.      Running pass 'Function Pass Manager' on module 'tmp.ll'.
2.      Running pass 'ARM Instruction Selection' on function '@fn'
 #0 0x0000598f54474272 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /data_nvme1n1/johbra01/llvm-project/llvm/lib/Support/Unix/Signals.inc:808:3
 #1 0x0000598f54471c0f llvm::sys::RunSignalHandlers() /data_nvme1n1/johbra01/llvm-project/llvm/lib/Support/Signals.cpp:105:20
 #2 0x0000598f54471fa9 SignalHandler(int, siginfo_t*, void*) /data_nvme1n1/johbra01/llvm-project/llvm/lib/Support/Unix/Signals.inc:418:13
 #3 0x000078b0eb845330 (/lib/x86_64-linux-gnu/libc.so.6+0x45330)
 #4 0x000078b0eb89eb2c __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x000078b0eb89eb2c __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x000078b0eb89eb2c pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x000078b0eb84527e raise ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x000078b0eb8288ff abort ./stdlib/abort.c:81:7
 #9 0x0000598f510b8142 llvm::json::operator==(llvm::json::Array const&, llvm::json::Array const&) /data_nvme1n1/johbra01/llvm-project/llvm/include/llvm/Support/JSON.h:580:58
#10 0x0000598f510b8142 llvm::json::operator==(llvm::json::Value const&, llvm::json::Value const&) (.cold) /data_nvme1n1/johbra01/llvm-project/llvm/lib/Support/JSON.cpp:206:28
#11 0x0000598f543ce185 (/data_nvme1n1/johbra01/llvm-project/build/bin/llc+0x3e3b185)
#12 0x0000598f54342420 llvm::DAGTypeLegalizer::SoftPromoteHalfResult(llvm::SDNode*, unsigned int) /data_nvme1n1/johbra01/llvm-project/llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp:32
86:58
#13 0x0000598f542c4fc0 llvm::DAGTypeLegalizer::run() /data_nvme1n1/johbra01/llvm-project/llvm/lib/CodeGen/SelectionDAG/LegalizeTypes.cpp:266:14
#14 0x0000598f542c6511 llvm::SmallVectorTemplateCommon<llvm::SDNode*, void>::isSmall() const /data_nvme1n1/johbra01/llvm-project/llvm/include/llvm/ADT/SmallVector.h:143:39
#15 0x0000598f542c6511 llvm::SmallVectorImpl<llvm::SDNode*>::~SmallVectorImpl() /data_nvme1n1/johbra01/llvm-project/llvm/include/llvm/ADT/SmallVector.h:603:23
#16 0x0000598f542c6511 llvm::SmallVector<llvm::SDNode*, 128u>::~SmallVector() /data_nvme1n1/johbra01/llvm-project/llvm/include/llvm/ADT/SmallVector.h:1203:3
#17 0x0000598f542c6511 llvm::DAGTypeLegalizer::~DAGTypeLegalizer() /data_nvme1n1/johbra01/llvm-project/llvm/lib/CodeGen/SelectionDAG/LegalizeTypes.h:31:31
#18 0x0000598f542c6511 llvm::SelectionDAG::LegalizeTypes() /data_nvme1n1/johbra01/llvm-project/llvm/lib/CodeGen/SelectionDAG/LegalizeTypes.cpp:1060:10
#19 0x0000598f54237157 llvm::TimeRegion::~TimeRegion() /data_nvme1n1/johbra01/llvm-project/llvm/include/llvm/Support/Timer.h:160:9
#20 0x0000598f54237157 llvm::NamedRegionTimer::~NamedRegionTimer() /data_nvme1n1/johbra01/llvm-project/llvm/include/llvm/Support/Timer.h:168:8
#21 0x0000598f54237157 llvm::SelectionDAGISel::CodeGenAndEmitDAG() /data_nvme1n1/johbra01/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:980:3
#22 0x0000598f5423bac5 llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) /data_nvme1n1/johbra01/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:1906:33
#23 0x0000598f5423d17a llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) /data_nvme1n1/johbra01/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:615:22
#24 0x0000598f51d92da8 (anonymous namespace)::ARMDAGToDAGISel::runOnMachineFunction(llvm::MachineFunction&) /data_nvme1n1/johbra01/llvm-project/llvm/lib/Target/ARM/ARMISelDAGToDAG.cpp:70:3
#25 0x0000598f54232c27 llvm::OptLevelChanger::~OptLevelChanger() /data_nvme1n1/johbra01/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:268:11
#26 0x0000598f54232c27 llvm::SelectionDAGISelLegacy::runOnMachineFunction(llvm::MachineFunction&) /data_nvme1n1/johbra01/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:375:
1
#27 0x0000598f54232c27 llvm::SelectionDAGISelLegacy::runOnMachineFunction(llvm::MachineFunction&) /data_nvme1n1/johbra01/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAGISel.cpp:346:
6
#28 0x0000598f53494a8a llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (.part.0) /data_nvme1n1/johbra01/llvm-project/llvm/lib/CodeGen/MachineFunctionPass.cpp:108:30
#29 0x0000598f53a5f984 llvm::FPPassManager::runOnFunction(llvm::Function&) /data_nvme1n1/johbra01/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1416:7
#30 0x0000598f53a5fba1 llvm::ilist_detail::node_base_prevnext<llvm::ilist_node_base<true, void>, true>::getNext() const /data_nvme1n1/johbra01/llvm-project/llvm/include/llvm/ADT/ilist_node_
base.h:42:38
#31 0x0000598f53a5fba1 llvm::ilist_node_impl<llvm::ilist_detail::node_options<llvm::Function, true, false, void, false, void>>::getNext() /data_nvme1n1/johbra01/llvm-project/llvm/include/ll
vm/ADT/ilist_node.h:119:66
#32 0x0000598f53a5fba1 llvm::ilist_iterator<llvm::ilist_detail::node_options<llvm::Function, true, false, void, false, void>, false, false>::operator++() /data_nvme1n1/johbra01/llvm-project
/llvm/include/llvm/ADT/ilist_iterator.h:187:25
#33 0x0000598f53a5fba1 llvm::FPPassManager::runOnModule(llvm::Module&) /data_nvme1n1/johbra01/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1451:22
#34 0x0000598f53a603ed runOnModule /data_nvme1n1/johbra01/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:1528:7
#35 0x0000598f53a603ed llvm::legacy::PassManagerImpl::run(llvm::Module&) /data_nvme1n1/johbra01/llvm-project/llvm/lib/IR/LegacyPassManager.cpp:539:55
#36 0x0000598f51203e66 compileModule(char**, llvm::LLVMContext&) /data_nvme1n1/johbra01/llvm-project/llvm/tools/llc/llc.cpp:755:34
#37 0x0000598f510bee77 main /data_nvme1n1/johbra01/llvm-project/llvm/tools/llc/llc.cpp:400:35
#38 0x000078b0eb82a1ca __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:74:3
#39 0x000078b0eb82a28b call_init ./csu/../csu/libc-start.c:128:20
#40 0x000078b0eb82a28b __libc_start_main ./csu/../csu/libc-start.c:347:5
#41 0x0000598f511f92e5 _start (/data_nvme1n1/johbra01/llvm-project/build/bin/llc+0xc662e5)
```

wzssyqa commented:
Thanks for your report. I am working on fix them.

mikaelholmen commented:
> Thanks for your report. I am working on fix them.

Maybe revert in the meantime? It's been broken for several days already.

wzssyqa commented:
It seems not hard to resolve. Let's wait for a while.
